libs/vkd3d/state.c

```diff
@@ -4663,21 +4670,42 @@ VkPipeline d3d12_pipeline_state_create_pipeline_variant(struct d3d12_pipeline_st
     pipeline_desc.pStages = graphics->stages;
     pipeline_desc.pViewportState = &vp_desc;
     pipeline_desc.pRasterizationState = &graphics->rs_desc;
-    pipeline_desc.pMultisampleState = &graphics->ms_desc;
     pipeline_desc.pDepthStencilState = &graphics->ds_desc;
-    pipeline_desc.pColorBlendState = &fragment_output_desc.cb_info;
     pipeline_desc.pDynamicState = &dynamic_create_info;
     pipeline_desc.layout = graphics->pipeline_layout;
     pipeline_desc.basePipelineIndex = -1;
 
     if (d3d12_device_supports_variable_shading_rate_tier_2(device))
         pipeline_desc.flags |= VK_PIPELINE_CREATE_RENDERING_FRAGMENT_SHADING_RATE_ATTACHMENT_BIT_KHR;
 
-    if (!(graphics->stage_flags & VK_SHADER_STAGE_MESH_BIT_EXT))
+    if (graphics->stage_flags & VK_SHADER_STAGE_TESSELLATION_CONTROL_BIT)
+        pipeline_desc.pTessellationState = &tessellation_info;
+
+    if (has_vertex_input_state)
     {
         pipeline_desc.pVertexInputState = &vertex_input_desc.vi_info;
         pipeline_desc.pInputAssemblyState = &vertex_input_desc.ia_info;
-        pipeline_desc.pTessellationState = &tessellation_info;
+    }
+
+    if (has_fragment_output_state || graphics->ms_desc.sampleShadingEnable)
+        pipeline_desc.pMultisampleState = &graphics->ms_desc;
+
+    if (has_fragment_output_state)
+        pipeline_desc.pColorBlendState = &fragment_output_desc.cb_info;
+
+    if (library_flags)
+    {
+        TRACE("Compiling pipeline library for %p with flags %#x.\n", state, library_flags);
+
+        library_create_info.sType = VK_STRUCTURE_TYPE_GRAPHICS_PIPELINE_LIBRARY_CREATE_INFO_EXT;
+        /* Explicit cast to silence a constness warning, this seems to be a Vulkan header bug */
+        library_create_info.pNext = (void*)pipeline_desc.pNext;
+        library_create_info.flags = library_flags;
+
+        pipeline_desc.pNext = &library_create_info;
+        pipeline_desc.flags |= VK_PIPELINE_CREATE_RETAIN_LINK_TIME_OPTIMIZATION_INFO_BIT_EXT;
```

Comment:
Missing VK_PIPELINE_CREATE_LIBRARY_BIT_KHR.

Reply:
Urg, why does validation not catch this.

Reply:
nvm I added this but the change is in another commit.